### ISSUE TITLE
Weird (incorrect?) watershed behavior

### ISSUE THREAD

**Issue by jni** [scikit-image/scikit-image#803]
Hi all,

This is something I might be inclined to tackle at some point, but I currently lack the bandwidth, so I thought I'd report it here. Essentially, it looks like the scikit-image 3D watershed function is dealing improperly with plateaus, though I could not reproduce this with some "toy" plateaus. Anyway, here's an EM image of some neurites:

![screen shot 2013-10-25 at 11 09 07 pm](https://f.cloud.github.com/assets/492549/1418042/536469ae-3fa7-11e3-8f48-3b740187e983.png)

And here's the (inverted) boundary probability map generated by a voxel-level classifier (using the cubehelix colormap):

![screen shot 2013-10-25 at 11 11 14 pm](https://f.cloud.github.com/assets/492549/1418049/82cd84c8-3fa7-11e3-8113-84f94b92282c.png)

You can see that the boundaries are rather thick. The probability is 1.0 across the whole boundary, ie it's a plateau. Here's the output from `skimage.morphology.watershed`:

![screen shot 2013-10-25 at 11 08 49 pm](https://f.cloud.github.com/assets/492549/1418861/751121ba-3fbf-11e3-91d0-654cdf6c7509.png)

You can see that one label has spread through the boundary, even though the "nearest" labels should propagate. Here's the output from my old pure python watershed [implementation](https://github.com/jni/gala/blob/master/gala/morpho.py#L242):

![screen shot 2013-10-25 at 11 08 31 pm](https://f.cloud.github.com/assets/492549/1418873/b9b92790-3fbf-11e3-902e-b4147ddc62e5.png)

This matches my expectation perfectly.

**Comment by kevin-keraudren** [comment]
Hi,

Here is a demo of before/after bug fix (in addition to the test I added in `scikit-image/skimage/morphology/tests`):

```python
import numpy as np
import matplotlib.pyplot as plt
import scipy.ndimage as nd

import skimage
from skimage.data import imread
from skimage.morphology import watershed
from skimage.color import rgb2gray

img = rgb2gray(imread("map.png"))

seg = imread("seg.png")

seeds, n_seeds = nd.label( nd.binary_erosion( img > 0.3, iterations=5 ) )

labels = watershed(-img, seeds)

fig, axes = plt.subplots(ncols=3, figsize=(8, 2.7))
ax0, ax1, ax2 = axes

ax0.imshow(img, cmap=plt.cm.gray, interpolation='nearest')
ax0.set_title('Map')
ax1.imshow(seeds, cmap=plt.cm.jet, interpolation='nearest')
ax1.set_title('Init')
ax2.imshow(labels, cmap=plt.cm.jet, interpolation='nearest')
ax2.set_title('Separated objects')

for ax in axes:
    ax.axis('off')

fig.subplots_adjust(hspace=0.01, wspace=0.01, top=1, bottom=0, left=0,
                    right=1)

plt.savefig('good.png')
```
Before:
![bad](https://cloud.githubusercontent.com/assets/802153/7730056/01d9e5da-ff10-11e4-8082-3a909ecdc66b.png)

After:
![good](https://cloud.githubusercontent.com/assets/802153/7730061/09af5cd6-ff10-11e4-9abb-7ce039678302.png)

**Comment by LeeKamentsky** [comment]
It looks as though that was my intent at some point. There's one more tweak - if you use a 3x3 structuring element, you want the diagonals to get processed last. Thanks for looking at this. I'll post links to the test and fix in CellProfiler

**Comment by LeeKamentsky** [comment]
I took a slightly different tack:
https://github.com/CellProfiler/CellProfiler/commit/b15c5facc37c9e06d7d469da691bb45edea34b63

The age is reset as you enter a plateau and is incremented as you advance into a plateau. I think this is closer to mimicing the FIFO of Vincent/Soille. There is a tweak I added that uses the distance of a structuring element from the structure's center as a minor qualifier for the age - this makes the algorithm consider horizontal or vertical neighbors before diagonal ones.

@kevin-keraudren , could you send me the map and init images in their raw form so I can test on mine?

**Comment by stefanv** [comment]
Also, let's agree on a common fix so that our two libraries are in sync.  That way, we can fix future bugs again in the same productive way.

**Comment by kevin-keraudren** [comment]
@LeeKamentsky I noticed you did not include a test for the new feature you added: In the test you added, you used a 3x3 cross structuring element. I would expect the output to be the same as before with a 3x3 cross, and the manhattan distance weighting would only play a role with structure elements starting from a 3x3 square. 

May I suggest the following test, which currently fails with 7209e1f33b1be0b05a0c03132094b8309af388e4 but should pass with your new implementation (if I understood correctly what you meant by breaking ties):

```python
    def test_watershed11(self):
        "watershed 11"
        data = np.array([[1, 1, 1],
                         [1, 1, 1]], np.uint8)
        markers = np.array([[1, 0, 0],
                            [0, 0, 2]], np.int8)
        out = watershed(data, markers, self.eight)
        print out
        error = diff([[1, 1, 2],
                      [1, 2, 2]], out)
        self.assertTrue(error < eps)
```

The test currently fails as it produces:
[[1 1 2]
 [1 1 2]]

However, if you use a cross element, it passes:
```python
        cross = np.array([[False, True, False],
                          [True, True, True],
                          [False, True, False]])
        out = watershed(data, markers, cross)
```

I guess the output of your new implementation will not always correspond to a 3x3 cross element, and breaking ties becomes more and more an issue for structure elements larger than 3x3.

Regarding the images map and init, you only need `map.png` (which comes from the first post on this bug report), as init is derived from it using a binary erosion (you can remove the line with `seg.png` in my example as it is not used): https://camo.githubusercontent.com/dc7d6e650a8f8da0c29de2c88cf2437fdc99e861/68747470733a2f2f662e636c6f75642e6769746875622e636f6d2f6173736574732f3439323534392f313431383034392f38326364383463382d336661372d313165332d383131332d3834663934623932323832632e706e67 
(right-click Save Image As)

**Comment by jni** [comment]
@kevin-keraudren @LeeKamentsky Glorious!!! Thank you! =D I'm over the moon. This thing has been nagging at me since I posted the issue! So simple yet so annoying! =P

@LeeKamentsky can you explain your solution a bit more, perhaps with a worked example? I don't understand why you need to keep track of plateaus — what matters is the relative age of points being added, and this should not change whether or not you reset upon enter? How do you detect entering a plateau anyway?

I also don't quite understand the arithmetic for your fancy ordered structuring element. =)

At any rate, thanks to both of you for tackling this!

**Comment by LeeKamentsky** [comment]
@kevin-keraudren - I'm afraid test 11 is going to fail unless we were to push the operations to set the pixels onto the priority queue (currently marker 1 sets all of the pixels of its neighbors, then marker 2 sets all of the pixels of its neighbors except for the ones set by marker 1). The algorithm could work like this:
* push all of the seeds on the queue
* loop (exit when queue is empty):
  * pop a location off of the queue
  * if the location is unmarked:
    * propagate the label to the location
    * push all unmarked neighbors onto queue with appropriate ages.

This uses more memory (can push the same destination location from several similarly-labeled pixels along the propagation front, requires storing the label in the priority queue) and is only marginally more accurate. I'm not sure if it's worth the added effort.
 
@jni - if you rely on the propagation age as I did, as opposed to the incremental age, a marker in the middle of a big basin will meet the edge of a plateau with an age that's larger than a marker in the middle of a small basin, so you need to reset as you enter a plateau. You know you're propagating across a plateau when the image value of the source pixel equals that of the destination pixel.
The ordering of the structuring element is based on the manhattan distance - the sum of the absolute values of the offsets. In 2-d, a diagonal has a value of 2 and horizontal/vertical have a value of 1. In 3-d, there's an additional diagonal to the corner which has a value of 3. I subtract 1 because the minimum distance is 1. I'm assuming nobody is going to use a structuring element that's more than 3 on a side... if that assumption is wrong, then you'd need to compute the euclidean distance, rank-order the unique values and assign the rank-order to the structuring elements. I had considered ordering the structuring elements by distance and then Kevin's age strategy might work if the horizontal/vertical elements entered the queue ahead of the diagonal ones.

**Comment by LeeKamentsky** [comment]
OK - last experiment until Tuesday. I ordered the structuring elements by Manhattan distance and then implemented both @kevin-keraudren 's age-by-insertion time and my age-by-propagation and ran the last test with an 8-connected structuring element. Both pass. I say Kevin's age-by-insertion is just plain simpler and easier for everyone to understand, so we should go with that is my vote.

Here's the commit:
https://github.com/LeeKamentsky/CellProfiler/commit/a244d3ca332da939ee12d6e6843206a915a0931a

**Comment by stefanv** [comment]
Thanks for the in-depth investigation, @LeeKamentsky.  @kevin-keraudren, I will leave the PR in your capable hands.

**Comment by kevin-keraudren** [comment]
Hi @stefanv , I understand from @LeeKamentsky 's last message that the PR is ready to be merged as it is. However the code in the master branch of CellProfiler now corresponds to the "age-by-propagation" approach, so you might prefer to wait until it is finalized so that both projects are kept in sync.

**Comment by jni** [comment]
Fixed by #1530
